### PR TITLE
Add benchmarking and actually improve timings

### PR DESCRIPTION
**Pull request by elbrujohalcon**
Thanks to @lauramcastro's dedication to the completeness of our paper, I actually benchmarked Hank and the results were… not what you would've expected…

# 😱 

|                               | Erlang/OTP   | Kazoo     | MongooseIM |
|-------------------------------|-------------:|----------:|-----------:|
|Size of project (in SLOC)      | `> 2M`       | `> 400K`  | `> 180K`   |
| | | | |
|Sequential File Parsing        | `29042ms`    | `3457ms`  | `1735ms`   |
|Parallel File Parsing          | **`8751ms`**     | **`959ms`**   | **`490ms`**    |
| | | | |
|Sequential Rule Evaluation     | **`150274ms`**   | **`6818ms`**  | **`2430ms`**   |
|Parallel Rule Evaluation       | `156136ms`   | `10678ms` | `2788ms`   |
| | | | |
|Total Sequential Runtime       | `181750ms`   | `10665ms` | `4372ms`   |
|Total Parallel Runtime         | `168057ms`   | `12101ms` | `3635ms`   |
| | | | |
|**Total Optimal Runtime**      | **`162424ms`**   | **`8360ms`**  | **`3393ms`**   |

In a nutshell, file parsing was greatly improved by using parallel execution, but rule evaluation was actually faster if performed sequentially. That may be unintuitive, but it can be explained by the fact that there are very few rules and none of them are bound by I/O or network processing. That's why the time spent by `rpc:pmap/3` spawning processes, waiting for results, and collating responses outweighs the time spent actually running the algorithms.

So, the _Optimal_ version you see in this PR removes parallelism from rule evaluation and gets us to the faster version of Hank ever.

### CODE REVIEW

**Comment by elbrujohalcon** [comment]
@pbrudnick / @diegomanuel I'm merging this one using admin powers… Feel free to let me know if you think I should adjust this code. I'll do it in future pull requests.

**Comment by pbrudnick** [comment]
I liked this @elbrujohalcon , I tested with similar results parallelizing inside some rules.
I only propose to maybe_get_stats if an option is passed to hank.

**Comment by elbrujohalcon** [comment]
You mean to avoid computing stats unless an flag is turned on? I considered it but it does make the code (and the tests) more complex for no real gain... I mean, calling `timer:tc/2` twice is never going to have a noticeable impact on the performance of the system.
And the stats are hidden from the end user unless `DIAGNOSTIC=1` anyway.

**Comment by pbrudnick** [comment]
Makes sense
